### DISCUSSION TITLE
Creating a queue for notifications

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#13976] [General]
I am trying to creating a messaging queue to allow scheduling notifications. The idea is sending notifications to users by allowing services to add message to send into a queue and then a background task will process these queue every minute. By default, I want to store the queue items into a database. At some point, I'll change the store to an implementation a cloud service.

Here is what I have done.

```c#
public class NotificationSenderBackgroundTask : IBackgroundTask
{
    private const string _lockerKey = "NotificationSenderBackgroundTask";

    public async Task DoWorkAsync(IServiceProvider serviceProvider, CancellationToken cancellationToken)
    {
        var notificationService = serviceProvider.GetService<INotificationService>();

        if (notificationService == null)
        {
            return;
        }

        // To prevent multiple background tasks running in parallel on multiple containers, lets create a lock.
        var distributedLock = serviceProvider.GetRequiredService<IDistributedLock>();

        var (locker, locked) = await distributedLock.TryAcquireLockAsync(_lockerKey, TimeSpan.FromMicroseconds(200));

        if (!locked)
        {
            // Could not place a lock. This means that another instance is already running.
            return;
        }

        await using var acquiredLock = locker;

        var store = serviceProvider.GetRequiredService<IDocumentManager<NotificationQueueDocument>>();

        var queueDocument = await store.GetOrCreateMutableAsync();

        if (queueDocument.Items == null || queueDocument.Items.Count == 0)
        {
            // Nothing to process.
            return;
        }

        var clock = serviceProvider.GetRequiredService<IClock>();

        var now = clock.UtcNow;

        // Get items from the queue that we should process.
        var items = queueDocument.Items.Where(i => i.Value.NextRunUtc < now).ToList();
        var usernames = new HashSet<string>();

        foreach (var item in items)
        {
            if (item.Value.Usernames == null)
            {
                // An item with no users should never been added to the queue. Let's remove it.
                queueDocument.Items.TryRemove(item.Key, out _);

                continue;
            }

            usernames.AddRange(item.Value.Usernames);
        }

        var session = serviceProvider.GetRequiredService<ISession>();
        var users = await session.Query<User, UserIndex>(i => i.IsEnabled && i.NormalizedUserName.IsIn(usernames)).ListAsync();

        foreach (var item in items)
        {
            if (item.Value.Usernames == null)
            {
                continue;
            }

            var enabledUsers = users.Where(x => item.Value.Usernames.Contains(x.NormalizedUserName)).ToList();

            foreach (var enabledUser in enabledUsers)
            {
                //TODO, call it in parallel.
                await notificationService.SendAsync(enabledUser, item.Value.Message);
            }

            if (!String.IsNullOrEmpty(item.Value.Expression))
            {
                var endAt = item.Value.EndAtUtc ?? DateTime.MaxValue;

                if (endAt > now)
                {
                    // At this point, we know that this item has a cron expression and could have a another instance.
                    // Let's try to figure out the next occurrence.
                    var schedule = CrontabSchedule.TryParse(item.Value.Expression);

                    if (schedule != null)
                    {
                        item.Value.LastRunUtc = now;
                        item.Value.NextRunUtc = schedule.GetNextOccurrence(now, endAt);

                        // Update the existing item on the queue.
                        queueDocument.Items[item.Key] = item.Value;

                        continue;
                    }
                }
            }

            // If we got this far, the item was processed and has no more occurrences.
            // We can safely remove it from the queue.
            queueDocument.Items.TryRemove(item.Key, out _);
        }

        // Update the queue.
        await store.UpdateAsync(queueDocument);
    }
}

public class NotificationQueueDocument : Document
{
    public ConcurrentDictionary<string, NotificationQueueItem> Items { get; set; } = new ConcurrentDictionary<string, NotificationQueueItem>();
}

public class NotificationQueueItem
{
    public string Expression { get; set; }

    public DateTime? EndAtUtc { get; set; }

    public DateTime? LastRunUtc { get; set; }

    public DateTime NextRunUtc { get; set; }

    public INotificationMessage Message { get; set; }

    /// <summary>
    /// The normalized username of the users to notify.
    /// </summary>
    public IList<string> Usernames { get; set; } = new List<string>();
}
```

### Question
When placing acquiring a lock, is there a way to manually unlock or removing it so the next time the background runs it can be picked up? Once the queue is updated, the lock can be removed.

Also, when using `IDocumentManager<NotificationQueueDocument>`, will I be able to update the queue from other services while the background task is running? For example, if the background task takes 2 minutes to finish, meanwhile other services use the document manager to add more items to the queue. Will the `store.UpdateAsync(queueDocument)` store all the updates made to the document from the background task or other services?

@jtkech I appreciate your feedback if you have an answer. Do you have better approach to handle this?

**Comment by jtkech** [comment]
There are so many scenarios, I would need to focus on it more, but here some thoughts.

 Yes we are missing sheduled tasks and jobs queue infrastructures, it is on my list but I may first try to complete the secrets module done by @deanmarcussen. So in the meantime you may need a custom implementation.

For any `IBackgroundTask` we can define a distributed lock but you may need to centralize it in a service that could not only process notifications but also regiter new ones from other services.

When you are `using` a distributed lock, as soon as your job is completed and that you go out the `using block`, the lock is automatically released (or after its expiration if the job is too long).

With `IDocumentManager` we ensure that all instances are in sync with the very last update, but the updates are not atomic (we have an `UpdateAtomicAsync()` but only for the `IVolatileDocumentManager`), so in itself it is not adapted for a job queue, but why not through a service using a dedicated distributed lock.

Also, if you may have a lot of notifications, maybe better to use a dedicated sql table, look at the `IndexingTaskManager`, how it deletes/adds tasks in the `IndexingTask` table, then how Lucene uses it and how it knows which items need to be updated. Hmm but here not the same, this is to update local indexes files, it uses a cursor (in a local file) on the `Id` column (auto incremented) of the last processed task (if multiple instances, not intended to share the same file system).

But in your case you may not need to have to implement a cursor, just a queue from a document or index table managed by a dedicated service using a dedicated distributed lock, this to process/add/delete jobs atomically. Humm, maybe before processing many jobs you could delete the entries first and release the lock so that other services can re-add jobs more quickly, hmm but if the process shut down some jobs may be lost.

The `NotificationQueueDocument` doesn't need a `ConcurrentDictionary`, on updating we should use an isolated instance and each component collaborating to the mutation, in a given scope, are not called concurrently.

**Reply by MikeAlhayek** [reply]
@jtkech Thanks for your feedback. I made lots of changes to allow for more flexible approach. Now the `QueueProcessorBackgroundTask` can process all queue items not specific to notifications. Just about anything you want to queue. Then using different implementations of `IQueueElementProcessor`, I can implement processor for different `QueueName`. In the code below, you'll see an implementation for processing notification based on content types `ContentNotificationQueueProcessor`. I user ContentPartHandler to create `QueueElement` when an item is published.

I published the code on a gist for you to view it if possible https://gist.github.com/MikeAlhayek/dd44628dcf649114bf9cf080074d7b62

I may create a store for `QueueElement` so that way if we want to store `QueueElement` in a different store than YesSql we are able to.

I appreciate your feedback.

### Updated

I implemented a store for storing the queue elements so we can change the store if needed. The default implementation `DefaultQueueStore` saves elements to the database.

**Reply by MikeAlhayek** [reply]
@jtkech I published this package on nuget just so I can test drive it in a project where I need to queue notification. It works as expected this far. I also published another use case [here ](https://www.nuget.org/packages/CrestApps.Queues/1.0.1) where one can use the same setup to queue up emails so they are send in the background.

**Reply by jtkech** [reply]
Okay cool, I'll try to look ai it this week end

**Reply by jtkech** [reply]
In the meantime, after a quick look I saw this

                var task = _notificationService.SendAsync(user, message);
            ...
            var totalSent = await Task.WhenAll(tasks);

Not recommended to use `Task.WhenAll()` because scoped services (so in a given scope) are not intended to be run concurrently, for example this is the case for `ISession`.

**Reply by MikeAlhayek** [reply]
@jtkech thank you. I update the gist with the new code which does not contained the `WhenAll` and updated the missing Background task,
